backend/src/apiserver/auth/auth.go

```diff
@@ -0,0 +1,44 @@
+// Copyright 2021 Arrikto Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+
+	"github.com/kubeflow/pipelines/backend/src/apiserver/client"
+	"github.com/kubeflow/pipelines/backend/src/apiserver/common"
+)
+
+type Authenticator interface {
+	GetUserIdentity(ctx context.Context) (string, error)
+}
+
+// Make this public for tests to force its re-instantiation
+var Authenticators []Authenticator
+
+func GetAuthenticators(tokenReviewClient client.TokenReviewInterface) []Authenticator {
+	if Authenticators == nil {
+		Authenticators = []Authenticator{
```

Comment:
I'm not very familiar with concurrent programming in golang. Can this be a race condition?
e.g. the server runs as many go routines, and there can be two incoming requests triggering this method at the same time.

/cc @capri-xiyue @elikatsis

Reply:
Do you have any concerns moving the init process to be part of server init?

Reply:
Hmm this could be an issue. Although I cannot think of a mesh of the threads' reads and writes that may result to bad values in `Authenticators` (the output is not related to the request, any request should result to the same authenticators being created), moving this to the initialization is probably a good idea.

But then we need to think what to do with this test (maybe drop it since testing the empty prefix shouldn't be part of the `auth_server` tests?): https://github.com/arrikto/kubeflow-pipelines/blob/3dc13dc773dcb240c8d15396b410051e221bfcb2/backend/src/apiserver/server/auth_server_test.go#L113

Reply:
Oh I almost forgot. The main concern is what to do with the token review client. The reason we are passing it to this function is to be able to use a fake one for tests. Is it possible to we have a workaround for this?

Reply:
I think a simple way to avoid using a global variable as cache is to move `GetAuthenticators()` call to `NewResourceManager`: https://github.com/kubeflow/pipelines/blob/5990285894d655b0b52ca4d9a2f08756a99f1ec8/backend/src/apiserver/resource/resource_manager.go#L99

and store the authenticators as properties of resource manager.

In this way, they are only initiated once, also you can test them in tests using the fake.

Reply:
the tricky thing about race condition is that, it's not only a matter of which order the two commands execute.
Both of them might be running at the same time and it can lead to memory corruption.

Although, I'm not sure whether memory corruption could happen when assigning an array to the same variable. In some tutorial the example was that when setting map key/values, two concurrent access can cause memory corruption.

Anyway, I'd prefer avoiding it, because it's not hard.

Reply:
I pushed a commit with this change and tested it a bit. It looks like it works :tada: 
PTAL :)